src/index.js

```diff
@@ -313,145 +289,174 @@ function addSegment(currentState, segment) {
+      if (segment instanceof EpsilonSegment) {
+        continue;
+      }
+
+      output += '/';
+      var segmentValue = segment.generate(params);
+      if (segmentValue === undefined) {
+        throw new Error(`A value is required for route parameter '${segment.name}' in route '${name}'.`);
+      }
 
-      output += "/";
-      output += segment.generate(params);
+      output += segmentValue;
     }
 
-    if (output.charAt(0) !== '/') { output = '/' + output; }
+    if (output.charAt(0) !== '/') {
+      output = '/' + output;
+    }
 
     if (params && params.queryParams) {
       output += this.generateQueryString(params.queryParams, route.handlers);
     }
 
     return output;
-  },
+  }
+
+  generateQueryString(params) {
+    var pairs = [], keys = [], encode = encodeURIComponent;
 
-  generateQueryString: function(params, handlers) {
-    var pairs = [];
-    var keys = [];
-    for(var key in params) {
+    for (var key in params) {
       if (params.hasOwnProperty(key)) {
         keys.push(key);
       }
     }
+
     keys.sort();
     for (var i = 0, len = keys.length; i < len; i++) {
       key = keys[i];
       var value = params[key];
-      if (value === null) {
+      if (value === null || value === undefined) {
         continue;
       }
-      var pair = encodeURIComponent(key);
-      if (isArray(value)) {
+
+      if (Array.isArray(value)) {
+        var arrayKey = `${encode(key)}[]`;
         for (var j = 0, l = value.length; j < l; j++) {
-          var arrayPair = key + '[]' + '=' + encodeURIComponent(value[j]);
-          pairs.push(arrayPair);
+          pairs.push(`${arrayKey}=${encode(value[j])}`);
         }
       } else {
-        pair += "=" + encodeURIComponent(value);
-        pairs.push(pair);
+        pairs.push(`${encode(key)}=${encode(value)}`);
       }
     }
 
-    if (pairs.length === 0) { return ''; }
+    if (pairs.length === 0) {
+      return '';
+    }
+
+    return '?' + pairs.join('&');
+  }
+
+  parseQueryString(queryString) {
+    var queryParams = {};
+    if (!queryString || typeof queryString !== 'string') {
+      return queryParams;
+    }
 
-    return "?" + pairs.join("&");
-  },
+    if (queryString.charAt(0) === '?') {
+      queryString = queryString.substr(1);
+    }
 
-  parseQueryString: function(queryString) {
-    var pairs = queryString.split("&"), queryParams = {};
-    for(var i=0; i < pairs.length; i++) {
-      var pair      = pairs[i].split('='),
-          key       = decodeURIComponent(pair[0]),
+    var pairs = queryString.split('&');
+    for(var i = 0; i < pairs.length; i++) {
+      var pair = pairs[i].split('='),
+          key = decodeURIComponent(pair[0]),
           keyLength = key.length,
           isArray = false,
           value;
-      if (pair.length === 1) {
-        value = 'true';
+
+      if (!key) {
+        continue;
+      } else if (pair.length === 1) {
+        value = true;
       } else {
         //Handle arrays
```

Comment:
We should consider also supporting the alternative array query string syntax, `?a=b&a=c`, when parsing.